lectures/zksnark.tex

```diff
@@ -0,0 +1,812 @@
+    $x_1$, the next set of points can be derived:
+    \begin{equation*}
+        \{(1,1), (2,0), (3,1), (4,-1)\}
+    \end{equation*}
+    We can see that it's used in the 1st, 3rd, and 4th constraints as the values of the coefficients
+    aren't zero.
+    
+    The Lagrange interpolation polynomial for this set of points can be built as follows:
+    \begin{align*}
+        \ell_1(x) &= \frac{(x - 2)(x - 3)(x - 4)}{(1 - 2)(1 - 3)(1 - 4)} = -\frac{(x - 2)(x - 3)(x - 4)}{6}, \\
+        \ell_2(x) &= \frac{(x - 1)(x - 3)(x - 4)}{(2 - 1)(2 - 3)(2 - 4)} = \frac{(x - 1)(x - 3)(x - 4)}{2}, \\
+        \ell_3(x) &= \frac{(x - 1)(x - 2)(x - 4)}{(3 - 1)(3 - 2)(3 - 4)} = -\frac{(x - 1)(x - 2)(x - 4)}{2}, \\
+        \ell_4(x) &= \frac{(x - 1)(x - 2)(x - 3)}{(4 - 1)(4 - 2)(4 - 3)} = \frac{(x - 1)(x - 2)(x - 3)}{6}.
+    \end{align*}
+    \begin{align*}
+        A_1(x) &= 1 \cdot \ell_1(x) + 0 \cdot \ell_2(x) + 1 \cdot \ell_3(x) + (-1) \cdot \ell_4(x) \\
+        &= -\frac{(x - 2)(x - 3)(x - 4)}{6} - \frac{(x - 1)(x - 2)(x - 4)}{2} - \frac{(x - 1)(x - 2)(x - 3)}{6} \\
+        &= -\frac{5}{6}x^3 + 6x^2 - \frac{79}{6}x + 9
+    \end{align*}
+    Therefore, the final Lagrange interpolation polynomial is:
+    \begin{equation*}
+        A_1(x) = -\frac{5}{6}x^3 + 6x^2 - \frac{79}{6}x + 9
+    \end{equation*}
+
+    As shown in Figure~\ref{fig:example-interpolation}, the curve intersects all the given points.
+    In this figure, the x-axis represents the constraint number, and the y-axis represents the 
+    coefficients of the $x_1$ пwitness element.
+\end{example}
+
+\begin{figure}[H]
+    \centering
+    \begin{tikzpicture}
+        \begin{axis}[
+            axis lines = middle,
+            xlabel = {$x$},
+            ylabel = {$A_1(x)$},
+            ymin = -2.99, ymax = 2.99,
+            xmin = -0.5, xmax = 4.99,
+            domain = 0:5,
+            samples = 100,
+            ytick = {-3,...,3},
+            xtick = {0,1,...,5},
+            grid = both, 
+            grid style = {line width=.1pt, draw=gray!20},
+            major grid style = {line width=.2pt, draw=gray!60}
+        ]
+        \addplot[
+            color=blue,
+            thick
+        ]
+        {-5/6*x^3 + 6*x^2 - 79/6*x + 9};
+        \addplot[
+            only marks,
+            mark=*,
+            color=red
+        ]
+        coordinates {(1,1) (2,0) (3,1) (4,-1)};
+
+        \node at (axis cs:1,1.5) [anchor=west] {\small (1,1)};
+        \node at (axis cs:2,0.2) [anchor=south] {\small (2,0)};
+        \node at (axis cs:3,1.5) [anchor=west] {\small (3,1)};
+        \node at (axis cs:4,-1) [anchor=west] {\small (4,-1)};
+        
+        \end{axis}
+    \end{tikzpicture}
+    \caption{The Lagrange inteprolation polynomial for points $\{(1,1), (2,0), (3,1), (4,-1)\}$}
+    \label{fig:example-interpolation}
+\end{figure}
+
+The degree of coefficient polynomials won't exceed $m - 1$.
```

Comment:
does not exceed

Reply:
Also, isn't it always exactly $m-1$?

Reply:
Yeah, fair enough. Even more trivial example is '{(1,1),(2,2),(3,3),(4,4)}', corresponding to 'f(x)=x'